HeartbeatLogging/Sources/RingBuffer.swift

```diff
@@ -0,0 +1,61 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+/// <#Description#>
+struct RingBuffer<Element>: Sequence {
+  /// An array of heartbeats treated as a circular queue and intialized with a fixed capacity.
+  private var circularQueue: [Element?]
+  /// The current "tail" and insert point for the `circularQueue`.
+  private var tailIndex: Int = 0
+
+  /// Intializes a `RingBuffer` with a given `capacity`.
+  /// - Parameter capacity: An `Int` representing the capacity.
+  init(capacity: Int) {
+    circularQueue = Array(repeating: nil, count: capacity)
+  }
+
+  /// Pushes an element to the back of the buffer, returning the element (`Element?`) that was overriten.
+  /// - Parameter element: The element to push to the back of the buffer.
+  /// - Returns: <#description#>
+  /// - Complexity: O(1)
+  @discardableResult
+  mutating func push(_ element: Element) -> Element? {
+    guard circularQueue.count > 0 else {
+      // Do not push if `circularQueue` is a fixed empty array.
+      return nil
+    }
+
+    defer {
+      // Increment index, wrapping around to the start if needed.
+      tailIndex += 1
+      tailIndex %= circularQueue.count
```

Comment:
What if `count` is 0?

Reply:
There is a guard statement at the top of the `push` API to ensure that count is greater than 0 by the time this block is reached.

Reply:
Just for my education, since I see there is a test for this - but I thought that defer blocks execute even after a guard? - see https://www.hackingwithswift.com/new-syntax-swift-2-defer

Reply:
Ah I see, good point. I think the difference here is that the `defer` block is defined after the `guard` statement. In the article above, the `defer` is defined before the `guard` statements.

Reply:
So when `count` is 0. The control flow exits in the guard before the defer block can ever be scheduled. 

To test it out, I ran the `testPush_WhenCapacityIsZero_DoesNothing` with a breakpoint in the defer block and the breakpoint was never triggered.